lib/LedgerNodeRecords.js

```diff
@@ -36,9 +37,36 @@ module.exports = class LedgerNodeRecords {
     // ...and if not, call it again and cache its result; if we don't have
     // that block height, we should throw `NotFoundError`
 
+    // use a cache for the latest block summary since this function may
+    // be called often
+    const getLastestBlockSummaryFn = () => this.ledgerNode.storage.blocks
+      .getLatestSummary(this.ledgerNode);
+    let latestBlockSummary = await this._latestBlockSummaryCache.memoize({
+      key: 'latestBlockSummary', fn: getLastestBlockSummaryFn});
+    let latestBlockHeight = latestBlockSummary.eventBlock.block.blockHeight;
+
+    // always set maxBlockHeight since it's used as part of an LRU key
+    if(maxBlockHeight === undefined) {
+      maxBlockHeight = latestBlockHeight;
+    } else if(maxBlockHeight > latestBlockHeight) {
+      // clear the summary cache if `maxBlockHeight` is greater than
+      // `latestBlockHeight` and set `maxBlockHeight` to freshest database value
+      // only if maxBlockHeight is still sless than latestBlockHeight
+      this._latestBlockSummaryCache.delete('latestBlockSummary');
+      latestBlockSummary = await this._latestBlockSummaryCache.memoize({
+        key: 'latestBlockSummary', fn: getLastestBlockSummaryFn});
+      latestBlockHeight = latestBlockSummary.eventBlock.block.blockHeight;
```

Comment:
@dlongley this duplicates code at the top of the function... I thought about breaking it out, but it felt like overkill. Any suggestions on how to simplify this code path to be more maintainable?

Reply:
A small helper called `_getLatestBlockHeight` that reads from the cache would do.

Reply:
Implemented in 18abd3d4a5c879464a57623054c5852d0ed6f00d.